### ISSUE TITLE
Bug: [strict-boolean-expressions] False positive "always false/true" with forEach()

### ISSUE THREAD

**Issue by m-akinc** [typescript-eslint/typescript-eslint#12036]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.3&fileType=.ts&code=MYGwhgzhAECyCeBhcVoG8CwAoa0AOATgJYBuYALgKbRkgCulA-AFzQD2ARgFaXDkDc2XLiH46HEEWDQAtvABidAHZ8ibJQAoAlKxJsiAE3Sjc5ABZEIAOloNoAXmjKDlAGZEllA4JzCA2gAMALpWrmwEAKJgwGYa2g4AfOjQ5pY2YPTUjmgAvvzQOVo%2BJtBErtAaqda2lFroOaINWDlAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs%2BSAIZN6AD2LQUySnyaoMkRNGh9okcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class MyClass {
  private value?: object;
    
  public myFunction(): void {
    this.value = undefined;
    [0].forEach(() => { this.value = {}; });

    if (this.value) {}
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/strict-boolean-expressions": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

There should be no error on `this.value` about the expression always being false, because it gets set to an object instance inside the `forEach` call.

### Actual Result

There is an error reported on `this.value`: "Unexpected nullish value in conditional. The condition is always false."

**Comment by kirkwaiblinger** [comment]
Hi @m-akinc!

This is a well-known limitation of the rule, so this will be a wontfix... We have it better documented under no-unnecessary-condition (see [link](https://typescript-eslint.io/rules/no-unnecessary-condition/#values-modified-within-function-calls)), but, since this has come up before, we also have a to-do item to extract it to a common location, see https://github.com/typescript-eslint/typescript-eslint/issues/11854. This issue report is validating for #11854 🙂 

FWIW, this is one of several reasons that I prefer to avoid `foreach` and instead always use `for...of` (which [doesn't suffer this defect](https://typescript-eslint.io/play/#ts=5.8.3&fileType=.ts&code=MYGwhgzhAECyCeBhcVoG8CwAoa0AOATgJYBuYALgKbRkgCulA-AFzQD2ARgFaXDkDc2XLiH46HEEWDQAtvABidAHZ8ibJQAoAlKxJsiAE3Sjc5ABZEIAOloNoAXmjKDlAGZEllA4JzDXbAmgNYHUIcmgAfXZXaABtAAYAXS1jX2Foc0sbMHpqRzQAXxNoIqxiohiNTOtbShTC0VKCoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs%2BSAIZN6AD2LQUySnyaoMkRNGh9okcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
